lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/evaluator/EvaluatorShim.java

```diff
@@ -0,0 +1,215 @@
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Evaluator shim class.
+ * This class sends a message back to the driver upon starting and listens for
+ * commands to launch an evaluator. Once the evaluator completes, it will listen
+ * for a command to terminate.
+ */
+@Private
+@EvaluatorSide
+public final class EvaluatorShim
+    implements EventHandler<RemoteMessage<EvaluatorShimProtocol.EvaluatorShimControlProto>> {
+  private static final Logger LOG = Logger.getLogger(EvaluatorShim.class.getName());
+
+  private final RemoteManager remoteManager;
+  private final REEFFileNames fileNames;
+  private final ConfigurationSerializer configurationSerializer;
+
+  private final String driverRemoteId;
+  private final String containerId;
+
+  private final EventHandler<EvaluatorShimProtocol.EvaluatorShimStatusProto> evaluatorShimStatusChannel;
+  private final AutoCloseable evaluatorShimCommandChannel;
+
+  private final ExecutorService evaluatorLaunchExecutorService;
+
+  private Process evaluatorProcess;
+  private Integer evaluatorProcessExitValue;
+
+  @Inject
+  EvaluatorShim(final REEFFileNames fileNames,
+                final ConfigurationSerializer configurationSerializer,
+                final RemoteManager remoteManager,
+                @Parameter(DriverRemoteIdentifier.class)
+                final String driverRemoteId,
+                @Parameter(ContainerIdentifier.class)
+                final String containerId) {
+    this.fileNames = fileNames;
+    this.configurationSerializer = configurationSerializer;
+
+    this.driverRemoteId = driverRemoteId;
+    this.containerId = containerId;
+
+    this.remoteManager = remoteManager;
+    this.evaluatorShimStatusChannel = this.remoteManager.getHandler(this.driverRemoteId,
+        EvaluatorShimProtocol.EvaluatorShimStatusProto.class);
+
+    this.evaluatorShimCommandChannel = this.remoteManager
+        .registerHandler(EvaluatorShimProtocol.EvaluatorShimControlProto.class, this);
+
+    this.evaluatorLaunchExecutorService = Executors.newSingleThreadExecutor();
+  }
+
+  public void run() {
+    LOG.log(Level.FINEST, "Entering EvaluatorShim.run().");
+    this.onStart();
+  }
+
+  public void stop() {
+    this.onStop();
+  }
+
+  @Override
+  public void onNext(final RemoteMessage<EvaluatorShimProtocol.EvaluatorShimControlProto> remoteMessage) {
+    final EvaluatorShimProtocol.EvaluatorShimCommand command = remoteMessage.getMessage().getCommand();
+    if (command == EvaluatorShimProtocol.EvaluatorShimCommand.EVALUATOR_LAUNCH) {
+      LOG.log(Level.INFO, "Received a command to launch the Evaluator.");
+      EvaluatorShim.this.onEvaluatorLaunch(remoteMessage.getMessage().getEvaluatorLaunchCommand(),
```

Comment:
Why use a class reference and not just "this.onEvaluatorLaunch"?

Reply:
One refactor too many :)